File: app/src/main/java/com/adsamcik/signalcollector/activities/LicenseActivity.java
package com.adsamcik.signalcollector.activities;

import android.os.Bundle;
import android.util.Log;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout;

import com.adsamcik.signalcollector.R;
import com.adsamcik.signalcollector.utility.Preferences;
import com.google.firebase.crash.FirebaseCrash;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;

import de.psdev.licensesdialog.LicensesDialog;
import de.psdev.licensesdialog.licenses.ApacheSoftwareLicense20;
import de.psdev.licensesdialog.licenses.License;
import de.psdev.licensesdialog.licenses.MITLicense;
import de.psdev.licensesdialog.model.Notice;

public class LicenseActivity extends DetailActivity {

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		setTheme(Preferences.getTheme(this));
		super.onCreate(savedInstanceState);
		LinearLayout parent = createScrollableContentParent(true);

		InputStream isMeta = getResources().openRawResource(R.raw.third_party_license_metadata);
		InputStream isLicense = getResources().openRawResource(R.raw.third_party_licenses);
		BufferedReader rMeta = new BufferedReader(new InputStreamReader(isMeta));
		BufferedReader rLicense = new BufferedReader(new InputStreamReader(isLicense));

		try {
			String name;
			while ((name = rMeta.readLine()) != null) {
				int firstSpace = name.indexOf(' ');
				int length = name.length();
				name = name.substring(firstSpace + 1, length);
				Button button = addButton(parent, name);
				addLicenseDialogListener(button, name, rLicense.readLine());
			}
		} catch (IOException e) {
			FirebaseCrash.report(e);
		}


		final String GSON = "Gson";
		addButton(parent, GSON).setOnClickListener(view -> {
			Notice notice = new Notice(GSON, "https://github.com/google/gson", "Copyright 2008 Google Inc", new ApacheSoftwareLicense20());
			new LicensesDialog.Builder(this)
					.setNotices(notice)
					.build()
					.show();
		});

		final String OKHTTP = "OkHttp";
		addButton(parent, OKHTTP).setOnClickListener(view -> {
			Notice notice = new Notice(OKHTTP, "https://github.com/square/okhttp", "Copyright 2016 Square, Inc.", new ApacheSoftwareLicense20());
			new LicensesDialog.Builder(this)
					.setNotices(notice)
					.build()
					.show();
		});

		final String LICENSE_DIALOG = "LicensesDialog";
		addButton(parent, LICENSE_DIALOG).setOnClickListener(view -> {
			Notice notice = new Notice(LICENSE_DIALOG, "https://github.com/PSDev/LicensesDialog", "Copyright 2013-2017 Philip Schiffer", new ApacheSoftwareLicense20());
			new LicensesDialog.Builder(this)
					.setNotices(notice)
					.build()
					.show();
		});


		setTitle(R.string.open_source_licenses);
	}

	private Button addButton(final ViewGroup parent, final String name) {
		Button button = new Button(this);
		button.setText(name);
		parent.addView(button);
		return button;
	}

	private void addLicenseDialogListener(Button button, String name, String licenseURL) {
		button.setOnClickListener(view -> {
			final Notice notice = resolveNotice(name, licenseURL);
			new LicensesDialog.Builder(this)
					.setNotices(notice)
					.build()
					.show();
		});
	}

	private Notice resolveNotice(String name, String licenseURL) {
		String lowerName = name.toLowerCase();
		if (lowerName.startsWith("stag")) {
			return new Notice(name, "https://github.com/vimeo/stag-java", "Copyright (c) 2016 Vimeo", new MITLicense());
		} else if (lowerName.equals("appintro")) {
			return new Notice("AppIntro", "https://github.com/apl-devs/AppIntro", "Copyright 2015 Paolo Rotolo\n" + "Copyright 2016 Maximilian Narr", new ApacheSoftwareLicense20());
		} else if (lowerName.equals("persistentcookiejar"))
			return new Notice("PersistentCookieJar", "https://github.com/franmontiel/PersistentCookieJar", "Copyright 2016 Francisco José Montiel Navarro", new ApacheSoftwareLicense20());
		else {
			final License license = resolveLicense(licenseURL);
			if (license == null)
				return new Notice(name, null, licenseURL, null);
			else
				return new Notice(name, null, null, license);
		}
	}

	private License resolveLicense(String url) {
		if (url.startsWith("http://www.apache.org/licenses/LICENSE-2.0") || url.startsWith("https://api.github.com/licenses/apache-2.0"))
			return new ApacheSoftwareLicense20();
		else if (url.startsWith("http://www.opensource.org/licenses/mit-license"))
			return new MITLicense();
		else
			return null;
	}
}

File: app/src/main/java/com/adsamcik/signalcollector/license/LicenseActivity.kt
package com.adsamcik.signalcollector.license

import android.content.Context
import android.content.pm.ActivityInfo
import android.graphics.Color
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.ViewGroup.LayoutParams.MATCH_PARENT
import android.widget.FrameLayout
import androidx.appcompat.widget.AppCompatButton
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.adsamcik.signalcollector.R
import com.adsamcik.signalcollector.common.Reporter
import com.adsamcik.signalcollector.common.activity.DetailActivity
import com.adsamcik.signalcollector.common.extension.dp
import com.adsamcik.signalcollector.common.style.IViewChange
import com.adsamcik.signalcollector.common.style.RecyclerStyleView
import com.adsamcik.signalcollector.common.style.StyleView
import de.psdev.licensesdialog.LicensesDialog
import de.psdev.licensesdialog.licenses.ApacheSoftwareLicense20
import de.psdev.licensesdialog.licenses.License
import de.psdev.licensesdialog.model.Notice
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader

/**
 * LicenseActivity shows users all the open source licenses for dependencies this app uses.
 * For some specific libraries there are manually filled in URLs and copyrights. For most licenses it uses Google's automatically generated list.
 */
class LicenseActivity : DetailActivity() {

	override fun onConfigure(configuration: Configuration) {
		configuration.elevation = 4.dp
		configuration.titleBarLayer = 1
		configuration.useColorControllerForContent = false
	}

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)

		val frameLayout = createFrameContentLayout(false)
		frameLayout.setBackgroundColor(Color.WHITE)

		val adapter = Adapter()
		val recycler = RecyclerView(this)
				.also { frameLayout.addView(it) }
				.apply {
					layoutParams = FrameLayout.LayoutParams(MATCH_PARENT, MATCH_PARENT)
					layoutManager = LinearLayoutManager(this@LicenseActivity).also { requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT }
					this.adapter = adapter
				}


		val licenseDialogString = "LicensesDialog"

		CustomLicense(licenseDialogString, Notice(licenseDialogString, "https://github.com/PSDev/LicensesDialog", "Copyright 2013-2017 Philip Schiffer", ApacheSoftwareLicense20()))
				.also { adapter.addLicense(it) }

		val materialCommunityIconTitle = "Material Design Community Icons"

		CustomLicense(materialCommunityIconTitle, Notice(materialCommunityIconTitle, "http://materialdesignicons.com/", "Copyright (c) 2014, Austin Andrews", object : License() {
			override fun getUrl(): String = "https://scripts.sil.org/cms/scripts/page.php?item_id=OFL_web"

			override fun getName(): String = "SIL OPEN FONT LICENSE Version 1.1"

			override fun readSummaryTextFromResources(context: Context?): String = getContent(context, R.raw.material_community_icon_license)

			override fun getVersion(): String = "1.1"
			override fun readFullTextFromResources(context: Context?): String = getContent(context, R.raw.material_community_icon_license)

		})).also { adapter.addLicense(it) }

		val resources = resources
		val isMeta = resources.openRawResource(R.raw.third_party_license_metadata)

		val rMeta = BufferedReader(InputStreamReader(isMeta))

		try {
			rMeta.forEachLine { line ->
				val split = line.split(' ')
				val fromTo = split[0].split(':')
				val license = ResourceLicenseObject(split[1], fromTo[0].toInt(), fromTo[1].toInt(), resources)
				adapter.addLicense(license)
			}
		} catch (e: IOException) {
			Reporter.report(e)
		}

		adapter.notifyDataSetChanged()

		setTitle(R.string.open_source_licenses)

		styleController.watchRecyclerView(RecyclerStyleView(recycler, 0))
		styleController.watchView(StyleView(frameLayout, 0, 0))
	}

	private class Adapter : RecyclerView.Adapter<Adapter.ViewHolder>(), IViewChange {
		private val licenses = mutableListOf<LicenseObject>()

		override var onViewChangedListener: ((View) -> Unit)? = null

		override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
			val root = LayoutInflater.from(parent.context).inflate(R.layout.layout_license_item, parent, false)
			onViewChangedListener?.invoke(root)
			return ViewHolder(root.findViewById(R.id.button), root)
		}

		override fun getItemCount(): Int = licenses.size

		override fun onBindViewHolder(holder: ViewHolder, position: Int) {
			val license = licenses[position]
			holder.button.text = license.name
			holder.button.tag = license
			holder.button.setOnClickListener(this::onButtonClicked)
		}

		private fun onButtonClicked(view: View) {
			showLicense(view.context, view.tag as LicenseObject)
		}

		private fun showLicense(context: Context, licenseObject: LicenseObject) {
			val notice = licenseObject.notice
			LicensesDialog.Builder(context)
					.setNotices(notice)
					.build()
					.show()
		}

		fun addLicense(license: LicenseObject) {
			licenses.add(license)
		}

		private class ViewHolder(val button: AppCompatButton, root: View) : RecyclerView.ViewHolder(root)
	}
}